### PR TITLE
Add an API for UCR data sources

### PR DESCRIPTION
**Pull request by czue**
## Product Description

Adds a new API that exposes UCR data source data (the JSON configurations, not the data itself). Including the ability to write back certain fields from the API.

## Technical Summary

Related to work that will be described better in https://github.com/dimagi/dashboard-oauth-poc/pull/1/

## Feature Flag

UCR.

### Safety story

It's a new API and it only exposes configuration objects, so it seems relatively safe, though the ability to modify UCR data sources does expose a fair number of possible risks (e.g. bypassing our existing validation). I haven't thought hard about how or whether these should be mitigated and plan to look a little closer to convince myself that the API does at a minimum the same thing the UI does.

### Automated test coverage

There aren't any tests. I can add them if it feels important, though it's not obviously high-value since at the moment the API is a pretty standard tastypie implementation.

### QA Plan

Not proposing any manual QA.

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by snopoke — COMMENTED** [review]
In principal this seems OK. You mention validation which is definitely something that should be added and one other thing is to ensure that the authorisation matches the current model.

**Comment by snopoke** [comment]
> > ensure that the authorisation matches the current model
> 
> Sorry can you say a bit more about what you mean by this?

Access to the API should require the same set of permissions that are required on the UI:
* FF enabled for user / domain
  * https://github.com/dimagi/commcare-hq/blob/d7795967fa8a2082425c1c26a3fcb0db8714021b/corehq/apps/userreports/views.py#L194
* User has the `edit_ucr` permission
  * https://github.com/dimagi/commcare-hq/blob/d7795967fa8a2082425c1c26a3fcb0db8714021b/corehq/apps/userreports/views.py#L220-L225

File: corehq/apps/api/resources/v0_5.py [hunk 1]
```diff
@@ -745,6 +747,74 @@ class Meta(CustomResourceMeta):
...
+    def _ensure_toggle_enabled(self, request):
+        if not toggles.USER_CONFIGURABLE_REPORTS.enabled_for_request(request):
+            raise ImmediateHttpResponse(HttpResponse(
+                json.dumps({"error": _("You don't have permission to access this API")}),
+                content_type="application/json",
+                status=401)
+            )
+
+    def obj_get(self, bundle, **kwargs):
+        self._ensure_toggle_enabled(bundle.request)
```
Comment (czue) [inline corehq/apps/api/resources/v0_5.py:768]: I considered a number of possible options before landing on "call this method at the top of all the other methods". 

1. Override `dispatch()` and run the checks first: Doesn't work because the authorization checks which populate the user and domain happen inside there. 
2. Override `dispatch()` and run the checks second: Doesn't work since the update method will already have already run its course.

The other options that may be better, but just didn't seem worth it include:

1. Add a new Authentication class that can handle both a permission check and a feature flag check (or more generically, a combinable authentication class)
2. Add a new `flag_enabled` decorator for APIs that can expect a `bundle` instead of a `request` - seemed a little specific and potentially overkill.

Open to different opinions or ideas I missed though.

**Comment by czue** [comment]
@snopoke I added checks and tested (manually) the conditions you mentioned. I think this is ready for review

File: corehq/apps/api/resources/v0_5.py (hunk 1 above)
Comment (snopoke) [reply corehq/apps/api/resources/v0_5.py:768]: Interesting, I see toggle checks and access to `request.user` in the dispatch method of `HqBaseResource`: https://github.com/dimagi/commcare-hq/blob/7c4b468e69f0bcdc174bad6f2388568f5ee3ef6e/corehq/apps/api/resources/__init__.py#L96-L109

Do those just not work then?

Fine to leave this as is.

**Review by snopoke — APPROVED** [review]

File: corehq/apps/api/resources/v0_5.py (hunk 1 above)
Comment (czue) [reply corehq/apps/api/resources/v0_5.py:768]: Yeah, it's a bit more complicated than I articulated (my comment was already long!).

Middleware runs prior to dispatch being called, so our `UsersMiddleware` will always populate an available domain, and if the user can be determined from the session then it will also be there: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/users/middleware.py#L37-L43

In the event of an alternative form of authentication - in the case I was testing, oauth, the user is populated via the `Authentication` class at this point in the superclass: https://github.com/django-tastypie/django-tastypie/blob/master/tastypie/resources.py#L495

Which eventually (in the case of oauth) finds its way to our decorator here: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/domain/decorators.py#L216

I didn't actually test API keys, but now I'm curious since it wasn't obvious at a glance where a user would be set.

**TL;DR: domain toggles should be fine. User toggles are more complicated around the edges and probably shouldn't be used in `dispatch`.**

Comment (czue) [reply corehq/apps/api/resources/v0_5.py:768]: aw my links didn't do the fancy thing your links did. :disappointed:

Comment (snopoke) [reply corehq/apps/api/resources/v0_5.py:768]: Thanks for the update, that's also what I surmised from a brief look at the code.

> aw my links didn't do the fancy thing your links did. disappointed

Interesting, I wonder why?

Comment (czue) [reply corehq/apps/api/resources/v0_5.py:770]: FYI, API keys also work (only when used this way). The user gets set here: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/domain/auth.py#L291-L292

Comment (czue) [reply corehq/apps/api/resources/v0_5.py:770]: > Interesting, I wonder why?

It looks like it doesn't expand the links if it's to a branch - presumably because it could change at any time.

**Comment by czue** [comment]
@snopoke I added one more commit with better validation / error handling: https://github.com/dimagi/commcare-hq/pull/31088/commits/2c063d619960943db590efee21bcda56579ab876

**Review by snopoke — APPROVED** [review]

**Comment by czue** [comment]
sorry found one more thing https://github.com/dimagi/commcare-hq/pull/31088/commits/e9b9d1bc41358474e5f2a958619b8cb26f8220bc

**Review by snopoke — APPROVED** [review]
